File: microservice-architecture/examples/microservice-basics/js/src/OrderService.js
import { Order } from './Order.js'

export class OrderService {
  constructor(inventoryClient) {
    this.inventoryClient = inventoryClient
  }

  createOrder(orderId, sku, quantity) {
    if (this.inventoryClient.reserve(sku, quantity)) {
      return new Order(orderId, sku, quantity, 'CREATED')
    }
    return new Order(orderId, sku, quantity, 'REJECTED')
  }
}

File: microservice-architecture/examples/microservice-basics/c/src/order_service.c
#include "func.h"

static Order create_order_impl(OrderService *service, const char *order_id, const char *sku, int quantity)
{
    Order order;
    strcpy(order.order_id, order_id);
    strcpy(order.sku, sku);
    order.quantity = quantity;

    if (service->inventory->reserve(service->inventory, sku, quantity)) {
        strcpy(order.status, "CREATED");
    } else {
        strcpy(order.status, "REJECTED");
    }

    return order;
}

void order_service_init(OrderService *service, InventoryService *inventory)
{
    service->inventory = inventory;
    service->create_order = create_order_impl;
}
